src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,296 @@
+		);
+	}
+
+	public async publish(
+		ctx: LogWriteContext,
+		logs: Array<DeviceLog & { version?: number }>,
+	): Promise<any> {
+		const streams = this.fromDeviceLogsToStreams(ctx, logs);
+		try {
+			let nextBackoff = MIN_BACKOFF;
+			let prevBackoff = MIN_BACKOFF;
+			while (nextBackoff <= MAX_BACKOFF) {
+				try {
+					return await this.push(ctx.belongs_to__application, streams);
+				} catch (error) {
+					// only retry when Loki is down or rate limiting
+					if (
+						[status.UNAVAILABLE, status.RESOURCE_EXHAUSTED].includes(error.code)
+					) {
+						this.ready = false;
+						await Bluebird.delay(nextBackoff);
+						// fibonacci backoff
+						nextBackoff = nextBackoff + prevBackoff;
+						prevBackoff = nextBackoff - prevBackoff;
+					} else {
+						captureException(error);
+						throw new BadRequestError(
+							`Failed to publish logs for device ${ctx.uuid}`,
+						);
+					}
+				}
+			}
+		} finally {
+			this.ready = true;
+		}
+	}
+
+	private push(appId: number, streams: StreamAdapter[]): Promise<any> {
+		const pushRequest = new PushRequest();
+		pushRequest.setStreamsList(streams);
+		return this.pusher.push(pushRequest, createOrgIdMetadata(String(appId)));
+	}
+
+	public subscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		if (!this.tailCalls.has(key)) {
+			const request = new TailRequest();
+			request.setQuery(this.getDeviceQuery(ctx));
+			request.setStart(createTimestampFromMilliseconds(Date.now()));
+
+			const call = this.querier.tail(
+				request,
+				createOrgIdMetadata(String(ctx.belongs_to__application)),
+			);
+			call.on('data', (response: TailResponse) => {
+				const stream = response.getStream();
+				if (stream) {
+					const logs = this.fromStreamToDeviceLogs(stream);
+					for (const log of logs) {
+						this.subscriptions.emit(key, log);
+					}
+				}
+			});
+			call.on('error', (err: Error & { details: string }) => {
+				if (err.details !== 'Cancelled') {
+					captureException(err, 'Loki tail call error.');
+				}
+				this.subscriptions.removeListener(key, subscription);
+				this.tailCalls.delete(key);
+			});
+			call.on('end', () => {
+				this.subscriptions.removeListener(key, subscription);
+				this.tailCalls.delete(key);
+			});
+			this.tailCalls.set(key, call);
+		}
+		this.subscriptions.on(key, subscription);
+	}
+
+	public unsubscribe(ctx: LogContext) {
+		const key = this.getKey(ctx);
+		const call = this.tailCalls.get(key);
+		call?.cancel();
+	}
+
+	private getDeviceQuery(ctx: LogContext) {
+		return `{device_id="${ctx.id}"}`;
+	}
+
+	private getKey(ctx: LogContext, suffix = 'logs') {
+		return `app:${ctx.belongs_to__application}:device:${ctx.id}:${suffix}`;
+	}
+
+	private getLabels(ctx: LogContext, log: DeviceLog): string {
+		return `{device_id="${ctx.id}", service_id="${
+			log.serviceId ? log.serviceId : 'null'
+		}", is_system="${log.isSystem}", is_std_err="${log.isStdErr}"}`;
+	}
+
+	private validateLog(log: DeviceLog) {
```

Comment:
You can mark this as an assertion function with something like:
```suggestion
	private validateLog(log: DeviceLog): asserts log is $type {
```

Reply:
Interesting did not know about this feature, updated.